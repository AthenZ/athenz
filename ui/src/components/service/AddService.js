/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import AddModal from '../modal/AddModal';
import AddServiceForm from './AddServiceForm';
import ServiceKeyUtils from '../utils/ServiceKeyUtils';
import RequestUtils from '../utils/RequestUtils';

export default class AddService extends React.Component {
    constructor(props) {
        super(props);
        this.api = this.props.api;
        this.onSubmit = this.onSubmit.bind(this);
        this.onChange = this.onChange.bind(this);
        this.state = {
            showModal: !!this.props.showAddService,
        };
    }

    onSubmit() {
        if (!this.state.name || this.state.name === '') {
            this.setState({
                errorMessage: 'Service name is required.',
            });
            return;
        }

        let keyValue = this.state.keyValue;

        if (keyValue && keyValue !== '') {
            keyValue = ServiceKeyUtils.y64Encode(
                ServiceKeyUtils.trimKey(keyValue)
            );
        }

        this.api
            .getService(this.props.domain, this.state.name)
            .then(() => {
                this.setState({
                    errorMessage:
                        'Service ' + this.state.name + ' already exists.',
                });
            })
            .catch((err) => {
                if (err.statusCode === 404) {
                    this.api
                        .addService(
                            this.props.domain,
                            this.state.name,
                            this.state.description,
                            this.state.providerEndpoint,
                            this.state.keyId,
                            keyValue,
                            this.props._csrf
                        )
                        .then(() => {
                            this.setState({ showModal: false });
                            this.props.onSubmit(`${this.state.name}`, false);
                        })
                        .catch((err) => {
                            this.setState({
                                errorMessage:
                                    RequestUtils.xhrErrorCheckHelper(err),
                            });
                        });
                } else {
                    this.setState({
                        errorMessage: RequestUtils.xhrErrorCheckHelper(err),
                    });
                }
            });
    }

    onChange(key, value) {
        this.setState({ [key]: value });
    }

    render() {
        return (
            <AddModal
                isOpen={this.state.showModal}
                cancel={this.props.onCancel}
                submit={this.onSubmit}
                title={`Add Service to ${this.props.domain}`}
                errorMessage={this.state.errorMessage}
                sections={
                    <AddServiceForm
                        api={this.api}
                        domain={this.props.domain}
                        onChange={this.onChange}
                        pageConfig={this.props.pageConfig}
                    />
                }
            />
        );
    }
}
