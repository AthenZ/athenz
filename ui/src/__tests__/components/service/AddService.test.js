/*
 * Copyright 2020 Verizon Media
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
import React from 'react';
import { render, fireEvent, waitForElement } from '@testing-library/react';
import AddService from '../../../components/service/AddService';
const pageConfig = {
    servicePageConfig: {
        keyCreationLink: {
            title: 'Key Creation',
            url:
                'https://test.com',
            target: '_blank',
        },
        keyCreationMessage:
            'Test Message',
    }
};
describe('AddService', () => {
    it('should render', () => {
        const cancel = function() {};
        const domain = 'domain';
        const api = {};
        const { getByTestId } = render(
            <AddService
                onCancel={cancel}
                domain={domain}
                api={api}
                showAddService={true}
                pageConfig={pageConfig}
            />
        );
        const addService = getByTestId('add-service-form');
        expect(addService).toMatchSnapshot();
    });

    it('should render error after addService throws error(refresh)', async () => {
        const cancel = function() {};
        const domain = 'domain';
        const api = {
            addService: function(
                domainName,
                serviceName,
                description,
                endpoint,
                keyId,
                keyValue,
                _csrf
            ) {
                return new Promise((resolve, reject) => {
                    reject({
                        statusCode: 0,
                    });
                });
            },
        };
        const {
            getByTestId,
            querySelector,
            getByText,
            getAllByTestId,
        } = render(
            <AddService
                onCancel={cancel}
                domain={domain}
                api={api}
                showAddService={true}
                pageConfig={pageConfig}
            />
        );
        const addServiceInput = await waitForElement(() =>
            getAllByTestId('input-node')
        );
        fireEvent.change(addServiceInput[0], {
            target: {
                value: 'test-name',
            },
        });
        fireEvent.change(addServiceInput[1], {
            target: {
                value: 'test-key',
            },
        });
        fireEvent.click(await waitForElement(() => getByText('Submit')));
        expect(
            await waitForElement(() => getByTestId('error-message'))
        ).toMatchSnapshot();
    });

    it('should render error after addService throws error(other)', async () => {
        const cancel = function() {};
        const domain = 'domain';
        const api = {
            addService: function(
                domainName,
                serviceName,
                description,
                endpoint,
                keyId,
                keyValue,
                _csrf
            ) {
                return new Promise((resolve, reject) => {
                    reject({
                        statusCode: 1,
                        body: {
                            message: 'test-error',
                        },
                    });
                });
            },
        };
        const {
            getByTestId,
            querySelector,
            getByText,
            getAllByTestId,
        } = render(
            <AddService
                onCancel={cancel}
                domain={domain}
                api={api}
                showAddService={true}
                pageConfig={pageConfig}
            />
        );
        const addServiceInput = await waitForElement(() =>
            getAllByTestId('input-node')
        );
        fireEvent.change(addServiceInput[0], {
            target: {
                value: 'test-name',
            },
        });
        fireEvent.change(addServiceInput[1], {
            target: {
                value: 'test-key',
            },
        });
        fireEvent.click(await waitForElement(() => getByText('Submit')));
        expect(
            await waitForElement(() => getByTestId('error-message'))
        ).toMatchSnapshot();
    });

    it('should call onSubmit if add succeeds', async () => {
        const cancel = function() {};
        const domain = 'domain';
        let test = 0;
        const onSubmit = function(succedMessage) {
            test = test + 1;
        };
        const api = {
            addService: function(
                domainName,
                serviceName,
                description,
                endpoint,
                keyId,
                keyValue,
                _csrf
            ) {
                return new Promise((resolve, reject) => {
                    resolve();
                });
            },
        };
        const {
            getByTestId,
            querySelector,
            getByText,
            getAllByTestId,
        } = render(
            <AddService
                onCancel={cancel}
                domain={domain}
                api={api}
                showAddService={true}
                onSubmit={onSubmit}
                pageConfig={pageConfig}
            />
        );
        const addServiceInput = await waitForElement(() =>
            getAllByTestId('input-node')
        );
        fireEvent.change(addServiceInput[0], {
            target: {
                value: 'test-name',
            },
        });
        fireEvent.change(addServiceInput[1], {
            target: {
                value: 'test-key',
            },
        });
        fireEvent.click(await waitForElement(() => getByText('Submit')));
        expect(await waitForElement(() => test)).toEqual(1);
    });

    it('should render error if no key name specified', async () => {
        const cancel = function() {};
        const domain = 'domain';
        const api = {
            addService: function(
                domainName,
                serviceName,
                description,
                endpoint,
                keyId,
                keyValue,
                _csrf
            ) {
                return new Promise((resolve, reject) => {
                    resolve();
                });
            },
        };
        const {
            getByTestId,
            querySelector,
            getByText,
            getAllByTestId,
        } = render(
            <AddService
                onCancel={cancel}
                domain={domain}
                api={api}
                showAddService={true}
                pageConfig={pageConfig}
            />
        );
        fireEvent.click(await waitForElement(() => getByText('Submit')));
        expect(
            await waitForElement(() => getByTestId('error-message'))
        ).toMatchSnapshot();
    });
});
