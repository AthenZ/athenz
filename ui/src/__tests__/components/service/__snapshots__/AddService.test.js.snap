// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddService should render 1`] = `
.emotion-0 {
  width: 100%;
  text-align: left;
  background-color: #fff;
}

.emotion-2 {
  -webkit-align-items: flex-start;
  -webkit-box-align: flex-start;
  -ms-flex-align: flex-start;
  align-items: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-flex-flow: row nowrap;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  padding: 10px 30px;
}

.emotion-4 {
  font-size: 14px;
  font-weight: 600;
  color: #303030;
  line-height: 36px;
  white-space: nowrap;
}

.emotion-5 {
  -webkit-flex: 0 0 100px;
  -ms-flex: 0 0 100px;
  flex: 0 0 100px;
  margin-right: 2%;
}

.emotion-7 {
  -webkit-flex: 1 1;
  -ms-flex: 1 1;
  flex: 1 1;
  margin-right: 10px;
}

.emotion-9 {
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-flex-flow: row nowrap;
  -webkit-flex-flow: row nowrap;
  -ms-flex-flow: row nowrap;
  flex-flow: row nowrap;
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  position: relative;
  width: 250px;
}

.emotion-9 input {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background: rgba(53,112,244,0.05);
  border-top: 2px solid transparent;
  border-bottom: 2px solid transparent;
  border-right: transparent;
  border-left: transparent;
  border-radius: 2px;
  box-shadow: none;
  box-sizing: border-box;
  color: #303030;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font: inherit;
  height: 36px;
  margin: 0;
  outline: none;
  padding: 0 1rem;
  text-align: left;
  width: 100%;
}

.emotion-9.animated input {
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
}

.emotion-9 input::-webkit-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-9 input::-moz-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-9 input:-ms-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-9 input::placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-9 input.focused,
.emotion-9 input:active,
.emotion-9 input:focus {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #3570f4;
  color: #303030;
}

.emotion-9 input:invalid {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #d01111;
  color: #303030;
}

.emotion-9 .message {
  font-size: 12px;
  left: 0;
  line-height: 1.8;
  position: absolute;
  top: 2.571rem;
}

.emotion-9.error input,
.emotion-9.error input.focused,
.emotion-9.error input:active,
.emotion-9.error input:focus {
  border-bottom: 2px solid #d01111;
}

.emotion-9.error .message {
  color: #d01111;
}

.emotion-10 {
  width: 500px;
}

.emotion-24 {
  -webkit-flex: 0 0 100px;
  -ms-flex: 0 0 100px;
  flex: 0 0 100px;
  margin-right: 2%;
  font-size: 14px;
  font-weight: 600;
  color: #303030;
  line-height: 36px;
  white-space: nowrap;
}

.emotion-28 {
  width: 500px;
}

.emotion-30 {
  color: #3570f4;
  -webkit-text-decoration: none;
  text-decoration: none;
  cursor: pointer;
}

.emotion-49 {
  font-family: Helvetica,Arial,sans-serif;
  font-size: 14px;
  font-weight: 300;
  box-sizing: border-box;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-flex-flow: column nowrap;
  -webkit-flex-flow: column nowrap;
  -ms-flex-flow: column nowrap;
  flex-flow: column nowrap;
  position: relative;
  width: 500px;
}

.emotion-49>textarea {
  -webkit-appearance: none;
  -moz-appearance: none;
  -ms-appearance: none;
  appearance: none;
  background: rgba(53,112,244,0.05);
  border-top: 2px solid transparent;
  border-bottom: 2px solid transparent;
  border-right: transparent;
  border-left: transparent;
  border-radius: 2px;
  box-shadow: none;
  box-sizing: border-box;
  color: #303030;
  -webkit-flex: 1 0 auto;
  -ms-flex: 1 0 auto;
  flex: 1 0 auto;
  font: inherit;
  height: auto;
  line-height: 1.5;
  margin: 0;
  outline: none;
  padding: 0.25em 0.75rem;
  text-align: left;
  -webkit-transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  transition: background-color 0.2s ease-in-out,color 0.2s ease-in-out,border 0.2s ease-in-out;
  width: auto;
}

.emotion-49>textarea::-webkit-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-49>textarea::-moz-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-49>textarea:-ms-input-placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-49>textarea::placeholder {
  color: rgba(48,48,48,0.6);
}

.emotion-49>textarea:active,
.emotion-49>textarea:focus {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #3570f4;
  color: #303030;
}

.emotion-49>textarea:invalid {
  background: rgba(53,112,244,0.05);
  border-bottom: 2px solid #d01111;
  color: #303030;
}

.emotion-49>textarea[disabled] {
  background: rgba(48,48,48,0.05);
  border-bottom: 2px solid rgba(48,48,48,0.05);
  color: rgba(48,48,48,0.25);
  cursor: not-allowed;
}

.emotion-49 .message {
  font-size: 12px;
  color: #303030;
  left: 0;
  line-height: 1.8;
  position: absolute;
  top: 2.571rem;
}

.emotion-50>textarea {
  resize: vertical;
}

<div
  autocomplete="off"
  class="emotion-0 emotion-1"
  data-testid="add-service-form"
>
  <div
    class="emotion-2 emotion-3"
  >
    <label
      class="emotion-4 denali-input-label emotion-5 emotion-6"
      for="service-name"
    >
      Service Name
    </label>
    <div
      class="emotion-7 emotion-8"
    >
      <div
        class="emotion-9 denali-input animated emotion-10 emotion-11"
        data-testid="input-wrapper"
      >
        <input
          autocomplete="off"
          data-testid="input-node"
          id="service-name"
          name="service-name"
          value=""
        />
        <div
          class="message"
          data-testid="message"
        />
      </div>
    </div>
  </div>
  <div
    class="emotion-2 emotion-3"
  >
    <label
      class="emotion-4 denali-input-label emotion-5 emotion-6"
      for="description"
    >
      Description
    </label>
    <div
      class="emotion-7 emotion-8"
    >
      <div
        class="emotion-9 denali-input animated emotion-10 emotion-11"
        data-testid="input-wrapper"
      >
        <input
          autocomplete="off"
          data-testid="input-node"
          id="description"
          name="description"
          value=""
        />
        <div
          class="message"
          data-testid="message"
        />
      </div>
    </div>
  </div>
  <div
    class="emotion-2 emotion-3"
  >
    <div
      class="emotion-24 emotion-25"
    >
      NOTE
    </div>
    <div
      class="emotion-7 emotion-8"
    >
      <div
        class="emotion-28 emotion-29"
      >
        Test Message
         
        Generate a
        <a
          class="emotion-30 emotion-31"
        >
           
          key pair
           
        </a>
        and add the public key below ONLY if you are not utilizing those.
      </div>
    </div>
  </div>
  <div
    class="emotion-2 emotion-33"
    data-testid="add-key-id"
  >
    <label
      class="emotion-4 denali-input-label emotion-5 emotion-36"
      for="key-id"
    >
      Public Key Id
    </label>
    <div
      class="emotion-7 emotion-38"
    >
      <div
        class="emotion-9 denali-input animated emotion-10 emotion-41"
        data-testid="input-wrapper"
      >
        <input
          autocomplete="off"
          data-testid="input-node"
          id="key-id"
          name="key-id"
          value=""
        />
        <div
          class="message"
          data-testid="message"
        />
      </div>
    </div>
  </div>
  <div
    class="emotion-2 emotion-33"
    data-testid="add-key-value"
  >
    <label
      class="emotion-4 denali-input-label emotion-5 emotion-36"
      for="key-value"
    >
      Key Value
    </label>
    <div
      class="emotion-7 emotion-38"
    >
      <div
        class="emotion-49 denali-textarea emotion-50 emotion-51"
        data-testid="textarea-wrapper"
      >
        <textarea
          data-testid="textarea"
          id="key-value"
          name="key-value"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`AddService should render error after addService throws error(other) 1`] = `
<div
  data-testid="error-message"
  name="red600"
  style="color: rgb(208, 17, 17); display: inline-block;"
>
  Status: 1. Message: test-error
</div>
`;

exports[`AddService should render error after addService throws error(refresh) 1`] = `
<div
  data-testid="error-message"
  name="red600"
  style="color: rgb(208, 17, 17); display: inline-block;"
>
  Session expired. Please refresh the page.
</div>
`;

exports[`AddService should render error if no key name specified 1`] = `
<div
  data-testid="error-message"
  name="red600"
  style="color: rgb(208, 17, 17); display: inline-block;"
>
  Service name is required.
</div>
`;

exports[`AddService should render error when after getService throws error(other) 1`] = `
<div
  data-testid="error-message"
  name="red600"
  style="color: rgb(208, 17, 17); display: inline-block;"
>
  Session expired. Please refresh the page.
</div>
`;

exports[`AddService should render error when service already exists 1`] = `
<div
  data-testid="error-message"
  name="red600"
  style="color: rgb(208, 17, 17); display: inline-block;"
>
  Status: 409. Message: Service test-name already exists
</div>
`;
